singularity-eos/eos/get_sg_eos_functors.hpp

```diff
@@ -216,12 +277,144 @@ struct final_functor {
     if (do_s) {
       sie_v(i) /= mass_sum;
     }
+    check_all_vals(i, npte, tid, pte_converged);
     /* reset mass fractions to original values if not normalized to 1 */
     for (int m = 0; m < nmat; ++m) {
       frac_mass_v(i, m) *= mass_sum;
     }
     return;
   }
+
+ private:
+  // Check to make sure returned values are normal or zero
+  template <typename valT>
+  PORTABLE_FORCEINLINE_FUNCTION int bad_val(valT const value,
+                                            const char *const var_name) const {
+    const bool good =
+        (value == valT{0} || (std::isnormal(1e10 * value) && std::isnormal(value)));
+    if (not good) {
+      using PortsOfCall::printf;
+      printf("### ERROR: Bad singularity-eos output\n  Var: %s\n  Value: %e\n", var_name,
+             value);
+    }
+    return not good;
+  }
+  // Check for prrecluding zero and negative (non-positive) values
+  template <typename valT>
+  PORTABLE_FORCEINLINE_FUNCTION int non_positive_val(valT const value,
+                                                     const char *const var_name) const {
+    const bool not_positive = value <= valT{0};
+    if (not_positive) {
+      using PortsOfCall::printf;
+      printf("### ERROR: Bad singularity-eos output\n  Var: %s\n  Value: %e\n", var_name,
+             value);
+    }
+    return not_positive;
+  }
+  // Check for prrecluding negative values
+  template <typename valT>
+  PORTABLE_FORCEINLINE_FUNCTION int negative_val(valT const value,
+                                                 const char *const var_name) const {
+    const bool negative = value < valT{0};
+    if (negative) {
+      using PortsOfCall::printf;
+      printf("### ERROR: Bad singularity-eos output\n  Var: %s\n  Value: %e\n", var_name,
+             value);
+    }
+    return negative;
+  }
+  // Only run these checks when built with debug flags
+  PORTABLE_FORCEINLINE_FUNCTION
+  void check_all_vals(int const i, int const npte, int const tid,
+                      bool const pte_converged) const {
+#ifndef NDEBUG
+    int n_bad_vals = 0;
+    for (int mp = 0; mp < npte; ++mp) {
```

Comment:
This is a broader style point than just these lines; I'm a fan of AAA - Almost Always Auto. `auto n_bad_vals = 0;` , `for (auto mp; mp < npte, ++mp)`, so on. 
If there isn't type restrictions due to passing values to Fortran (where this style is probably bad), I'd consider AAA as the preferred C++ style.

Reply:
I'm not sure I'm 100% bought in to AAA, but I can see how it could be helpful in these two places. The index is pretty easy to reason about since it is constrained either by the indexing behavior of the containers or the mapping `pte_mats()`. I don't anticipate the compiler doing anything weird with the type here.

For `n_bad_vals`, I suppose the code works with both a boolean type or an integer type so I'm being too constrictive. I was just worried that maybe there was some aspect of boolean math I didn't understand that might bite me (for example if true doesn't necessarily have to be 1). So I specifically chose an `int` here so that it was easier to reason about what exactly was going to happen at each step (I did at least make sure that the `bool` -> `int` conversion was guaranteed to be defined).

If I'm being overly cautious though, I can be persuaded to change both to be auto.

Reply:
I'm going to push back on @mauneyc-LANL here. Types are a form of documentation IMO and I lean *against* auto unless it explicitly generalizes the code.

Regarding bools vs ints I actually think all this logic about bad values should be boolean. I don't see why we need ints at all here. I make a comment elsewhere in the review describing what I would do as far as that is concerned.

Reply:
I think I split the difference here. I made index variables auto since their type is easily deduced from the Kokkos views or the iteration extent.

I want to keep the type of the bad values counter (now a `bool` and `any_bad_vals`) since I feel that it communicates a lot about the behavior of that variable in the function.